logstash-core-queue-jruby/src/main/java/org/logstash/ackedqueue/ext/JrubyAckedQueueMemoryExtLibrary.java

```diff
@@ -0,0 +1,141 @@
+    // are assumed to be logstash Event.
+
+
+    @JRubyClass(name = "AckedMemoryQueue", parent = "Object")
+    public static class RubyAckedMemoryQueue extends RubyObject {
+        private Queue queue;
+
+        public RubyAckedMemoryQueue(Ruby runtime, RubyClass klass) {
+            super(runtime, klass);
+        }
+
+        public Queue getQueue() {
+            return this.queue;
+        }
+
+        // def initialize
+        @JRubyMethod(name = "initialize", optional = 3)
+        public IRubyObject ruby_initialize(ThreadContext context, IRubyObject[] args)
+        {
+            args = Arity.scanArgs(context.runtime, args, 3, 0);
+
+            int capacity = RubyFixnum.num2int(args[1]);
+            int maxUnread = RubyFixnum.num2int(args[2]);
+
+            Settings s = new MemorySettings(args[0].asJavaString());
+            PageIOFactory pageIOFactory = (pageNum, size, path) -> new ByteBufferPageIO(pageNum, size, path);
+            CheckpointIOFactory checkpointIOFactory = (source) -> new MemoryCheckpointIO(source);
+            s.setCapacity(capacity);
+            s.setMaxUnread(maxUnread);
+            s.setElementIOFactory(pageIOFactory);
+            s.setCheckpointIOFactory(checkpointIOFactory);
+            s.setElementClass(Event.class);
+
+            this.queue = new Queue(s);
+
+            return context.nil;
+        }
+
+        @JRubyMethod(name = "open")
+        public IRubyObject ruby_open(ThreadContext context)
+        {
+            try {
+                this.queue.getCheckpointIO().purge();
+                this.queue.open();
+            } catch (IOException e) {
+                throw context.runtime.newIOErrorFromException(e);
+            }
+
+            return context.nil;
+        }
+
+        @JRubyMethod(name = {"write", "<<"}, required = 1)
+        public IRubyObject ruby_write(ThreadContext context, IRubyObject event)
+        {
+            if (!(event instanceof JrubyEventExtLibrary.RubyEvent)) {
+                throw context.runtime.newTypeError("wrong argument type " + event.getMetaClass() + " (expected LogStash::Event)");
+            }
+
+            long seqNum;
+            try {
+                seqNum = this.queue.write(((JrubyEventExtLibrary.RubyEvent) event).getEvent());
+            } catch (IOException e) {
+                throw context.runtime.newIOErrorFromException(e);
+            }
+
+            return context.runtime.newFixnum(seqNum);
+        }
+
+        @JRubyMethod(name = "read_batch", required = 2)
+        public IRubyObject ruby_read_batch(ThreadContext context, IRubyObject limit, IRubyObject timeout)
+        {
+            Batch b;
+
+            try {
+                b = this.queue.readBatch(RubyFixnum.num2int(limit), RubyFixnum.num2int(timeout));
+            } catch (IOException e) {
+                throw context.runtime.newIOErrorFromException(e);
+            }
+
+            // TODO: return proper Batch object
```

Comment:
When the Ruby based Read/Write Client and Batch code is stable I want to merge it into JRuby Extn. IIRC this is a TODO to return a ReadBatch or similar because ATM we have a `ReadBatch(RubyAckedBatch(Batch))`